### DISCUSSION TITLE
Mirth 4.5.1 HL7 and keeping '&'

### DISCUSSION THREAD

**Discussion by AcuteParanoia** [nextgenhealthcare/connect#6295] [Q&A] [answered]
I am receiving this as part of my HL7: OBX|4|TX|500-1104976|660FLDGRP4|^^^^^^^^^^^19800907^5000000010V924669-L^C0011^MICROVASIVE & STENT||||||F
which Mirth does get.  In my source transformer I have a Mapper type step with msg['OBX'][3]['OBX.5']['OBX.5.15'].toString().trim(); but I am only getting 'MICROVASIVE' and then a failed to write to SQL.  I need to be able to save the full message 'MICROVASIVE & STENT'.  I figure out it doesn't like '&' and '~' in the incoming message.  How can I preserve those characters?

**Comment by pacmano1** [comment]
Tell your sender to properly escape HL7 delimiters as per spec.  

Read https://docs.intersystems.com/latest/csp/docbook/DocBook.UI.Page.cls?KEY=EHL72_escape_sequences

**Reply by AcuteParanoia** [reply]
I agree but unfortunately that's not an option.

**Reply by pacmano1** [reply]
maybe search the forums first?  https://forums.mirthproject.io/forum/mirth-connect/support/11948-removing-from-a-message-subsegment

**Reply by AcuteParanoia** [reply]
Remember I'm actually trying to preserve it in the segment, not remove it.  In fact in the source transformer I don't even see it, so something has happened to it prior to that point.  I do have the data types set to HL7 v2.x

**Comment by joaryche** [comment, marked as answer]
As you may know in MSH.2 the separator characters are located here where

the default delimiters are vertical bar or pipe (|) for the field separator, caret (^) for the component separator, ampersand (&) for the subcomponent separator, and number sign (#) for the default truncation separator. The tilde (~) is the default repetition separator.

MSH|^~\&|

1. You could change out the subcomponent delimiter to something else so the ampersand in your OBX would not be counted.

2. You could search you incoming message to either replace or escape the ampersands in that field.

3. You could go into the channel's Data Type Wizard for HL7v2.x and check off the "Parse Subcomponents" option on the Serialization (incoming) Group.

**Reply by pacmano1** [reply]
1 might break a lot
2 ok, but the same exercise as Nick's suggestion but you gotta find the right regex for it.
3 probably not what most people are use to.

**Reply by AcuteParanoia** [reply]
It appears that 3 worked in a quick test.  Is there something else this could break (i.e. leaving Parse Subcomponents unchecked)?

**Reply by pacmano1** [reply]
There is nothing wrong with what @joaryche mentioned on bullet 3 - assuming you don't care about other subcomponents anywhere else in the message.

In other words, it your message and how you parse and ingest it is your call .

**Reply by AcuteParanoia** [reply]
I have a 183 variables that I setup in the source transformer that each just get their piece and then it's thru a destination filter and a straight save to a SQL table.  Nothing really amazing.

Thanks for your help!

**Reply by AcuteParanoia** [reply]
Oddly enough '&' works but '~' still doesn't, but in my case the '&' is more common.

**Reply by pacmano1** [reply]
There are usually smarter ways to do that rather than set all those variables.

For example, you can just create a single variable of JSON and insert this to your database.   

```JS
var stuff = 
{ 
  "mrn" : msg.....,
  "lastname" : msg....
}
```
Then in your db engine create a stored procedure or function that accepts that JSON and inserts/updates as needed.

Obviously you might be collecting patient, visit, etc information.  Even arrays of information.   The "tricky" part is getting the SQL correct for the insert/update depending on the engine.

A very distinct advantage of this is you have a single parameter for the paramterized SQL call should be doing.

e.g. in postgres:

```JS
var sql = <>
select sp_upsert_data(?::jsonb)
</>.toString();
var params = new java.util.ArrayList();
params.add(stuff)
/// call the database
```

**Reply by AcuteParanoia** [reply]
Thanks for that info.  I will have to play with it some since I've never really worked with JSON, but I understand what you are saying.
